modules/core/src/main/java/org/apache/ignite/internal/processors/cache/consistentcut/ConsistentCut.java

```diff
@@ -0,0 +1,260 @@
+    private volatile Set<IgniteInternalFuture<IgniteInternalTx>> removedActive = ConcurrentHashMap.newKeySet();
+
+    /** */
+    ConsistentCut(GridCacheSharedContext<?, ?> cctx, UUID id) {
+        this.cctx = cctx;
+        this.id = id;
+
+        log = cctx.logger(ConsistentCut.class);
+    }
+
+    /** */
+    public UUID id() {
+        return id;
+    }
+
+    /**
+     * Inits local Consistent Cut: prepares list of active transactions to check which side of Consistent Cut they belong to.
+     */
+    protected void init() throws IgniteCheckedException {
+        walLog(new ConsistentCutStartRecord(id), false);
+
+        beforeCut = ConcurrentHashMap.newKeySet();
+        afterCut = ConcurrentHashMap.newKeySet();
+
+        GridCompoundFuture<Boolean, Boolean> checkFut = new GridCompoundFuture<>(CU.boolReducer());
+
+        Iterator<IgniteInternalFuture<IgniteInternalTx>> finFutIt = cctx.tm().activeTransactions().stream()
+            .filter(tx -> tx.state() != ACTIVE)
+            .map(IgniteInternalTx::finishFuture)
+            .iterator();
+
+        // Invoke sequentially over two iterators:
+        // 1. iterators are weakly consistent.
+        // 2. we need a guarantee to handle `removedActive` after `activeTxs` to avoid missed transactions.
+        checkTransactions(finFutIt, checkFut);
+        checkTransactions(removedActive.iterator(), checkFut);
+
+        removedActive = null;
+
+        checkFut.markInitialized();
+
+        checkFut.listen(finish -> {
+            if (Boolean.FALSE.equals(finish.result()) || isDone()) {
+                if (log.isDebugEnabled())
+                    log.debug("Cut might be inconsistent for id " + id + ". Skip writing FinishRecord.");
+
+                onDone(new IgniteCheckedException("Cut is inconsistent."));
+
+                return;
+            }
+
+            try {
+                WALPointer ptr = walLog(new ConsistentCutFinishRecord(id, beforeCut, afterCut), true);
+
+                onDone(ptr);
+            }
+            catch (IgniteCheckedException e) {
+                U.error(log, "Failed to write ConsistentCutFinishRecord to WAL for id " + id, e);
+
+                onDone(e);
+            }
+        });
+    }
+
+    /**
+     * Collects a transaction before it is removed from {@link IgniteTxManager#activeTransactions()}.
+     *
+     * @param txFinFut Transaction finish future.
+     */
+    public void onRemoveActiveTransaction(IgniteInternalFuture<IgniteInternalTx> txFinFut) {
+        Set<IgniteInternalFuture<IgniteInternalTx>> txs = removedActive;
+
+        if (txs != null)
+            txs.add(txFinFut);
+    }
+
+    /**
+     * Checks active transactions - decides which side of Consistent Cut they belong to after they finished.
+     *
+     * @param activeTxFinFuts Collection of active transactions to check.
+     * @param checkFut Compound future that reduces finishes of checked transactions.
+     */
+    private void checkTransactions(
```

Comment:
This function can be combined with the `checkFut.listen` callback from line 116.
No need to continue check transaactions in case we found something errouness and can just cancel cut.

Reply:
This is implemented within the `boolReducer` that is used for creating the compound future. It stops collecting futures results after any of them was finished with `false`.